api/jobs/topEvents.js

```diff
@@ -0,0 +1,98 @@
+                                if (typeof result.map[mapKey] !== "object" && result.map[mapKey]) {
+                                    eventMap.push(result.list[index]);
+                                }
+                            }
+                            else {
+                                eventMap.push(result.list[index]);
+                            }
+                        }
+                        if (Array.isArray(eventMap)) {
+                            const arrPeriod = ["30days", "hour"];
+                            const appID = String(app_id.toString());
+                            const collectionName = "top_events" + crypto.createHash("sha1").update(appID).digest("hex");
+                            common.db.collection(collectionName).drop();
+                            async.eachSeries(arrPeriod, function(givenPeriod, eachCallback) {
+                                var data = {};
+                                var ob = {
+                                    app_id,
+                                    appTimezone,
+                                    qstring: {
+                                        period: givenPeriod
+                                    }
+                                };
+                                setTimeout(function() {
+                                    async.each(eventMap, function(event, doneEventMap) {
+                                        var collectionNameEvents = "events" + crypto.createHash("sha1").update(event + app_id).digest("hex");
+                                        countlyApi.data.fetch.getTimeObjForEvents(collectionNameEvents, ob, function(doc) {
+                                            countlyEvents.setDb(doc || {});
+                                            var countProp = countlyEvents.getNumber("c");
+                                            data[event] = {};
+                                            data[event].data = {
+                                                count: countProp
+                                            };
+                                            doneEventMap();
+                                        });
+                                    },
+                                    function() {
+                                        if (givenPeriod === "hour") {
+                                            givenPeriod = "today";
+                                        }
+                                        common.db.collection(collectionName).insert({
+                                            ts: Math.round(new Date().getTime() / 1000),
+                                            period: givenPeriod,
+                                            data
+                                        });
+                                    }
+                                    );
+                                    eachCallback();
+                                }, 5000);
```

Comment:
Setting timeout in such way, is usually a bad design. Most probably you needed to do all of this in the callback of collection drop?